libsolidity/analysis/FunctionCallGraph.cpp

```diff
@@ -0,0 +1,207 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libsolidity/analysis/FunctionCallGraph.h>
+
+using namespace std;
+using namespace solidity::frontend;
+
+bool FunctionCallGraphBuilder::CompareByID::operator()(Node const& _lhs, Node const& _rhs) const
+{
+	if (_lhs.index() != _rhs.index())
+		return _lhs.index() < _rhs.index();
+
+	if (std::holds_alternative<SpecialNode>(_lhs))
+		return std::get<SpecialNode>(_lhs) < std::get<SpecialNode>(_rhs);
+	return std::get<ASTNode const*>(_lhs)->id() < std::get<ASTNode const*>(_rhs)->id();
+}
+
+bool FunctionCallGraphBuilder::CompareByID::operator()(Node const& _lhs, int64_t _rhs) const
+{
+	solAssert(!std::holds_alternative<SpecialNode>(_lhs), "");
+
+	return std::get<ASTNode const*>(_lhs)->id() < _rhs;
+}
+
+bool FunctionCallGraphBuilder::CompareByID::operator()(int64_t _lhs, Node const& _rhs) const
+{
+	solAssert(!std::holds_alternative<SpecialNode>(_rhs), "");
+
+	return _lhs < std::get<ASTNode const*>(_rhs)->id();
+}
+
+shared_ptr<FunctionCallGraphBuilder::ContractCallGraph> FunctionCallGraphBuilder::create(ContractDefinition const& _contract)
+{
+	m_contract = &_contract;
+
+	m_graph = make_shared<ContractCallGraph>(_contract);
+
+	// Create graph for constructor, state vars, etc
+	m_currentNode = SpecialNode::CreationRoot;
+	m_currentDispatch = SpecialNode::CreationDispatch;
+	for (ContractDefinition const* contract: _contract.annotation().linearizedBaseContracts)
+		visitConstructor(*contract);
```

Comment:
Does this create the right edges - i.e. from one constructor to the next?

Reply:
No, but we can't do that properly anyway, what if there are no explicit constructors somewhere in the inheritance chain?

Reply:
Can you briefly mention this in the header?

Reply:
Why exactly are missing explicit constructors in the inheritance chain a problem here? You can just connect to the next existing one, can't you? We can of course get this merged with open issues like that, but I'm afraid things like this will become a problem, once we want to move on with something like https://github.com/ethereum/solidity/pull/10143, if that's still the plan...

Reply:
Sure, but wouldn't that be terrible inconsistent? How can you work with that in code if you don't know if that constructor now is the next one or if there was a missing one in between .. lot's of extra checks required to take advantage of that properly...

Reply:
I'm not quite sure I understand the issue. It'd just mean removing ``visitConstructor`` (which is only ever called from here) and doing it right here instead, while keeping ``m_currentNode`` up to date wouldn't it? Maybe in reversed order, but apart from that...
```
for (ContractDefinition const* contract: _contract.annotation().linearizedBaseContracts | boost::adaptors::reversed)
{
	for (auto const* stateVar: _contract.stateVariables())
		stateVar->accept(*this);

	for (auto arg: _contract.baseContracts())
		arg->accept(*this);

	if (_contract.constructor())
	{
		add(*m_currentNode, _contract.constructor());
		_contract.constructor()->accept(*this);
		m_currentNode = _contract.constructor();
	}
}
```

Reply:
But yeah, I may be missing something and it's not exactly crucial to get this done in this very PR - but we might need it before follow-up PRs, if we want to avoid weird behaviour.